### ISSUE TITLE
[Slim 4 - Error handling] PHP Script does not stop running when error happens with custom ErrorHandlerMiddleware

### ISSUE THREAD

**Issue by samuelgfeller** [odan/slim4-tutorial#11]
I followed the [error handling](https://odan.github.io/2020/05/27/slim4-error-handling.html) article to log PHP errors in a custom log file. 
I ran faulty code like `$array = []; $test = $array['nada'];` but the error is skipped and client gets a 200 code. Error is however logged in the right place.
When I remove the middleware, the code breaks (which is good), client request fails and error is logged in default error.log. 
Should I manually throw an exception [like here](https://stackoverflow.com/questions/10520390/stop-script-execution-upon-notice-warning)? 
I seem to be not understanding a crucial part.  

Also when error is thrown (except if I catch the error manually), CORS header are not added to the response (I use the same CORS Middleware as in your blog) and client blocks the response, so I get no error code 500 and no message. It's 
`(failed)net::ERR_FAILED | xhr`. 
Is there a way to still have a proper response for the client? 
`middleware.php` looks like this:
```php
return function (App $app) {
    $container = $app->getContainer();
    $logger = $container->get(LoggerInterface::class);

    $app->add(JwtClaimMiddleware::class);

    $app->add(CorsMiddleware::class);
    $app->add(JsonBodyParserMiddleware::class);
    $app->addRoutingMiddleware();
    
    // If uncommented error not thrown (question 1)
    // $app->add(ErrorHandlerMiddleware::class); 
    $app->addErrorMiddleware(true, true, true, $logger);
};
```

**Comment by odan** [comment]
Hi @samuelgfeller

Question 1:

> Should I manually throw an exception like here?

If you also want to abort the execution in this particular case, you can throw an `ErrorException` or even a more specific exception. Just customize the `ErrorHandlerMiddleware::process` method according to your requirements.

Question 2:

>  CORS header are not added to the response

The `CorsMiddleware` is an outgoing middleware and doesn't catch any Exceptions. So the Slim `ErrorHandler` will catch the exception and creates a new Response object without the CORS header. The problem is that the Slim `ErrorHandler::responde` method creates a new Response object. So all your response CORS headers will be lost anyway.

I try to find a solution.

**Comment by odan** [comment]
I think I found a workaround to ensure that the CORS headers are always added by modifying the `CorsMiddleware` and adding a middleware to forward the response.

File: `src/Middlware/CorsMiddleware.php`:

```php
<?php

namespace App\Middleware;

use App\Exception\CorsMiddlewareException;
use Exception;
use Psr\Http\Message\ResponseFactoryInterface;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use Slim\Routing\RouteContext;

/**
 * CORS middleware.
 */
final class CorsMiddleware implements MiddlewareInterface
{
    /**
     * @var ResponseFactoryInterface
     */
    private $responseFactory;

    /**
     * The constructor.
     *
     * @param ResponseFactoryInterface $responseFactory The response factory
     */
    public function __construct(ResponseFactoryInterface $responseFactory)
    {
        $this->responseFactory = $responseFactory;
    }

    /**
     * Invoke middleware.
     *
     * @param ServerRequestInterface $request The request
     * @param RequestHandlerInterface $handler The handler
     *
     * @return ResponseInterface The response
     */
    public function process(
        ServerRequestInterface $request,
        RequestHandlerInterface $handler
    ): ResponseInterface {
        $routeContext = RouteContext::fromRequest($request);
        $routingResults = $routeContext->getRoutingResults();
        $methods = $routingResults->getAllowedMethods();
        $requestHeaders = $request->getHeaderLine('Access-Control-Request-Headers');

        try {
            $response = $handler->handle($request);
        } catch (Exception $exception) {
        }

        if (!isset($response)) {
            $response = $this->responseFactory->createResponse(500);
        }

        $response = $response
            ->withHeader('Access-Control-Allow-Origin', '*')
            ->withHeader('Access-Control-Allow-Methods', implode(', ', $methods))
            ->withHeader('Access-Control-Allow-Headers', $requestHeaders ?: '*');

        // Optional: Allow Ajax CORS requests with Authorization header
        $response = $response->withHeader('Access-Control-Allow-Credentials', 'true');

        if (isset($exception)) {
            $response = $response->withHeader('Content-Type', 'application/json');

            // Add custom response body here...
            $response->getBody()->write(json_encode([
                'error' => [
                    'message' => $exception->getMessage(),
                ]
            ]));

            // Throw exception to pass the response with the CORS headers
            throw new CorsMiddlewareException($response, $exception->getMessage(), 500, $exception);
        }

        return $response;
    }
}
```

File: `src/Exception/CorsMiddlewareException.php`:

```php
<?php

namespace App\Exception;

use Psr\Http\Message\ResponseInterface;
use RuntimeException;
use Throwable;

final class CorsMiddlewareException extends RuntimeException
{
    /**
     * @var ResponseInterface
     */
    private $response;

    public function __construct(
        ResponseInterface $response,
        string $message,
        int $code = 500,
        Throwable $previous = null
    ) {
        parent::__construct($message, $code, $previous);

        $this->response = $response;
    }

    public function getResponse(): ResponseInterface
    {
        return $this->response;
    }
}
```

File: `src/Middleware/CorsMiddlwareExceptionMiddleware.php`

```php
<?php

namespace App\Middleware;

use App\Exception\CorsMiddlewareException;
use App\Factory\LoggerFactory;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use Psr\Log\LoggerInterface;

final class CorsMiddlwareExceptionMiddleware implements MiddlewareInterface
{
    /**
     * @var LoggerInterface
     */
    private $logger;

    /**
     * The constructor.
     *
     * @param LoggerFactory $loggerFactory The logger
     */
    public function __construct(LoggerFactory $loggerFactory)
    {
        $this->logger = $loggerFactory
            ->addFileHandler('errors.log')
            ->createInstance('cors_error_middleware');
    }

    /**
     * Invoke middleware.
     *
     * @param ServerRequestInterface $request The request
     * @param RequestHandlerInterface $handler The handler
     *
     * @return ResponseInterface The response
     */
    public function process(
        ServerRequestInterface $request,
        RequestHandlerInterface $handler
    ): ResponseInterface {
        try {
            return $handler->handle($request);
        } catch (CorsMiddlewareException $exception) {
            // Logging
            $this->logger->error($exception->getMessage());

            // Return the CORS response
            return $exception->getResponse();
        }
    }
}
```

Add the new middleware in `config/middleware.php`

```php
<?php

use App\Middleware\CorsMiddleware;
use App\Middleware\ErrorHandlerMiddleware;
use App\Middleware\CorsMiddlwareExceptionMiddleware;
use Selective\BasePath\BasePathMiddleware;
use Slim\App;
use Slim\Middleware\ErrorMiddleware;

return function (App $app) {xml
    $app->addBodyParsingMiddleware();
    $app->add(CorsMiddleware::class);
    $app->addRoutingMiddleware();
    $app->add(BasePathMiddleware::class);

    $app->add(CorsMiddlwareExceptionMiddleware::class); // <--- Add this entry

    $app->add(ErrorHandlerMiddleware::class);
    $app->add(ErrorMiddleware::class);
};
```

The container.php requires these entries:

```php
use App\Factory\LoggerFactory;
use Psr\Container\ContainerInterface;
use Psr\Http\Message\ResponseFactoryInterface;
// ...

return [
    // ...

    LoggerFactory::class => function (ContainerInterface $container) {
        return new LoggerFactory($container->get('settings')['logger']);
    },

    ResponseFactoryInterface::class => function (ContainerInterface $container) {
        return $container->get(App::class)->getResponseFactory();
    },
};
```

**Comment by samuelgfeller** [comment]
Question 1  
> If you also want to abort the execution in this particular case, you can throw an ErrorException or even a more specific exception. Just customize the ErrorHandlerMiddleware::process method according to your requirements.

Right. Somehow I thought the execution had to stop when the array key doesn't exist but it's only a notice... So the behaviour is normal. Facepalm on me for this one.

Question 2  
That is awesome, thank you very much! I definitely wouldn't have come up with this. 

Unfortunately there is still something little that makes it not work on my code.   
First I got `Cannot create RouteContext before routing has been completed` so I moved `$app->addRoutingMiddleware();` below `$app->add(CorsMiddleware::class);`. But then although the error went away, headers were not set.   
I debugged the whole thing and found out that the error is thrown in `CorsMiddleware.php` and it adds the wanted headers to the request and throws `CorsMiddlewareException`.   
But then it goes into the slim `ErrorMiddleware.php` where the following code is:
```php
try {
    return $handler->handle($request);
} catch (Throwable $e) {
    return $this->handleException($request, $e);
}
```
Exception caught again and headers replaced. So I tried removing `$app->addErrorMiddleware(true, true, true, $logger);`. 
`middleware.php`
```php
$app->add(JwtClaimMiddleware::class);

$app->add(JsonBodyParserMiddleware::class);

$app->add(CorsMiddleware::class);
$app->addRoutingMiddleware();

$app->add(ErrorHandlerMiddleware::class);
// $app->addErrorMiddleware(true, true, true, $logger);
```
But now after going into `CorsMiddlewareException.php` and executing `$this->response = $response` xdebug doesn't take me anywhere else and response is sent - without headers.

**Comment by odan** [comment]
It's important to add the `CorsMiddlwareExceptionMiddleware` to the middleware stack into the right position.

```php
use App\Middleware\CorsMiddlwareExceptionMiddleware;

// ...

 $app->add(CorsMiddlwareExceptionMiddleware::class); // <--- Add this entry
```

**Comment by samuelgfeller** [comment]
What is the content of `CorsMiddlwareExceptionMiddleware`? 
Since you showed `src/Middlware/CorsMiddleware.php` content with exception handling I thought it was just a typo in `middleware.php` and that you meant `CorsMiddleware.php` (with new exception handling), my bad.

**Comment by samuelgfeller** [comment]
Would it make sense to open an issue on Slim repo and ask that `addErrorMiddleware()`
does not replace the present headers but keeps the old ones ? Not creating a new response if response is already present.

**Comment by odan** [comment]
> What is the content of CorsMiddlwareExceptionMiddleware?

See my code example from above.

> Would it make sense to open an issue on Slim repo and ask that addErrorMiddleware()
does not replace the present headers but keeps the old ones ? Not creating a new response if response is already present.

I was thinking exactly the same thing. The question would be, how could this be achieved without breaking changes?

Have you added the `CorsMiddlwareExceptionMiddleware`?

**Comment by samuelgfeller** [comment]
> Have you added the `CorsMiddlwareExceptionMiddleware`  

Thanks for the update and yes now after I've added it, IT WORKS! 

> I was thinking exactly the same thing. The question would be, how could this be achieved without (breaking changes)?

Indeed. I certainly can not (yet) determine that and come up with a suggestion. Most I could do is create an issue but even that wouldn't be accurate and probably said in bad terms and not straight on the point.    
If you don't mind and think that it's a worthy addition to slim may I kindly ask you to bring it up there?   

Maybe not worthy enough though I've seen practically nothing on the internet about that (perhaps I searched wrong terms) so I feel like kind of the only person that faces this issue...   
And the solution you provided is well well enough.
